core/src/apps/cardano/auxiliary_data.py

```diff
@@ -3,118 +3,171 @@
+    addresses.validate_address_parameters(address_parameters)
 
-    if not SCHEMA_STAKING_ANY_ACCOUNT.match(
-        catalyst_registration_parameters.staking_path
-    ):
-        raise wire.ProcessError("Invalid auxiliary data")
+    if parameters.voting_purpose is not None:
+        assert_cond(parameters.format == CardanoGovernanceRegistrationFormat.CIP36)
 
-    address_parameters = catalyst_registration_parameters.reward_address_parameters
-    if address_parameters.address_type == CardanoAddressType.BYRON:
-        raise wire.ProcessError("Invalid auxiliary data")
 
-    addresses.validate_address_parameters(address_parameters)
+def _validate_voting_public_key(key: bytes) -> None:
+    assert_cond(len(key) == GOVERNANCE_VOTING_PUBLIC_KEY_LENGTH)
+
+
+def _validate_delegations(
+    delegations: list[messages.CardanoGovernanceDelegation],
+) -> None:
+    assert_cond(len(delegations) <= MAX_DELEGATION_COUNT)
+    for delegation in delegations:
+        _validate_voting_public_key(delegation.voting_public_key)
+
+
+def _get_voting_purpose_to_serialize(
+    parameters: messages.CardanoGovernanceRegistrationParametersType,
+) -> int | None:
+    if parameters.format == CardanoGovernanceRegistrationFormat.CIP15:
+        return None
+    if parameters.voting_purpose is None:
+        return DEFAULT_VOTING_PURPOSE
+    return parameters.voting_purpose
 
 
 async def show(
     ctx: wire.Context,
     keychain: seed.Keychain,
     auxiliary_data_hash: bytes,
-    catalyst_registration_parameters: messages.CardanoCatalystRegistrationParametersType
-    | None,
+    parameters: messages.CardanoGovernanceRegistrationParametersType | None,
     protocol_magic: int,
     network_id: int,
     should_show_details: bool,
 ) -> None:
-    if catalyst_registration_parameters:
-        await _show_catalyst_registration(
+    if parameters:
+        await _show_governance_registration(
             ctx,
             keychain,
-            catalyst_registration_parameters,
+            parameters,
             protocol_magic,
             network_id,
+            should_show_details,
         )
 
     if should_show_details:
-        await show_auxiliary_data_hash(ctx, auxiliary_data_hash)
+        await layout.show_auxiliary_data_hash(ctx, auxiliary_data_hash)
 
 
-async def _show_catalyst_registration(
+async def _show_governance_registration(
     ctx: wire.Context,
     keychain: seed.Keychain,
-    catalyst_registration_parameters: messages.CardanoCatalystRegistrationParametersType,
+    parameters: messages.CardanoGovernanceRegistrationParametersType,
     protocol_magic: int,
     network_id: int,
+    should_show_details: bool,
 ) -> None:
-    public_key = catalyst_registration_parameters.voting_public_key
-    encoded_public_key = bech32.encode(bech32.HRP_JORMUN_PUBLIC_KEY, public_key)
-    staking_path = catalyst_registration_parameters.staking_path
+    for delegation in parameters.delegations:
```

Comment:
We probably need to display some msg in the UI for empty delegation arrays.

Reply:
As resolved elsewhere, empty delegation arrays will remain forbidden.